### PR TITLE
Socket.io server and client implementation

### LINKED ISSUE #578: Contract query returning aborted tx contract info
The sql query to get deployed contract metadata isn't factor in when a contract was deployed multiple times (1 successfully, others unsuccessfully). The postgres query planner then non-deterministically decides which to return.

This results in errors where the aborted contract tx metadata is used for materializing `contract-call` tx args.

For example, the contract `SP32AEEF6WW5Y0NMJ1S8SBSZDAY8R5J32NBZFPKKZ.wrapped-nothing-v6` was deployed first here:
* https://explorer.stacks.co/txid/0x13786cf0e6c40d58c47d51284163ea80b6e3dbbae23a2fc070412b77ffe00dd6?chain=mainnet then here:
* https://explorer.stacks.co/txid/SP32AEEF6WW5Y0NMJ1S8SBSZDAY8R5J32NBZFPKKZ.wrapped-nothing-v6?chain=mainnet

In one of the API deployments, postgres is returning the aborted entry, and causing the error:
```
GET /extended/v1/tx/0x1abf5404717dc7432683842bd0e068276e57164c1c52c7c02fcdb6d36ab0fbd8
{
"error": "TypeError: Cannot read property 'functions' of null",
"stack": "TypeError: Cannot read property 'functions' of null\n    at Object.getTxFromDataStore (/root/stacks-blockchain-api/src/api/controllers/db-controller.ts:781:39)\n    at runMicrotasks (<anonymous>)\n    at processTicksAndRejections (internal/process/task_queues.js:93:5)\n    at /root/stacks-blockchain-api/src/api/routes/tx.ts:196:21\n    at wrappedMiddleware (/root/stacks-blockchain-api/node_modules/@awaitjs/express/index.js:62:7)",
"errorTag": "21101b50-30e2-45d0-903f-1fc1d6b9f30d"
}
```

### PR DESCRIPTION
**Pull request by zone117x**
Implements a socket.io server which provides an easy way for clients to subscribe to new mempool transactions, blocks, transactions for a given address, and STX balance updates for a given address.

See https://github.com/blockstack/rx for a client library which uses the socket.io client to create RxJS observables. 

A small socket.io client helper was implemented in the `@stacks/blockchain-api-client`, but it's also easy to use with the regular `socket.io-client` lib:
```ts
import { io } from 'socket.io-client';

const socket = io('https://stacks-node-api.testnet.stacks.co/', {
  query: { subscriptions: 'mempool' }
});

socket.on('mempool', tx => console.log('mempool', tx));
```

The socket.io event interface (shared between both server and client): 
https://github.com/blockstack/stacks-blockchain-api/blob/97811d627ab811fe30fed4efb677dd8f26282b46/docs/socket-io/index.d.ts#L1-L23

* Also fixes #578
* Implements a new endpoint: `/extended/v1/tx/block_height/{height}`

### CODE REVIEW

File: docs/openapi.yaml [hunk 1]
```diff
@@ -2178,8 +2181,45 @@ paths:
               example:
                 $ref: ./api/transaction/get-transactions.example.json
 
+  /extended/v1/tx/block_height/{height}:
+    get:
+      operationId: get_transactions_by_block_height
+      summary: Transactions in a block
+      description: Get all transactions in block
```
Comment (kyranjamie) [inline docs/openapi.yaml:2188]:
```suggestion
      description: Get all transactions in block by height
```

File: src/tests/api-tests.ts [hunk 2]
```diff
@@ -3890,23 +3890,43 @@ describe('api tests', () => {
...
+    );
+    expect(result4.body.limit).toBe(20);
+    expect(result4.body.offset).toBe(15);
+    expect(result4.body.total).toBe(1);
+    expect(result4.body.results.length).toBe(0);
   });
 
   afterEach(async () => {
-    await new Promise<void>(resolve => api.server.close(() => resolve()));
+    await api.terminate();
```
Comment (kyranjamie) [inline src/tests/api-tests.ts:3929]: image

File: docs/socket-io/index.d.ts [hunk 3]
```diff
@@ -0,0 +1,23 @@
+import type { AddressStxBalanceResponse, AddressTransactionWithTransfers, Block, MempoolTransaction } from '..';
```
Comment (kyranjamie) [inline docs/socket-io/index.d.ts:1]: is this included in the types package?

**Review by kyranjamie — APPROVED** [review]

File: docs/socket-io/index.d.ts (hunk 3 above)
Comment (zone117x) [reply docs/socket-io/index.d.ts:1]: Yeah, this _is_ the types package 👍

**Review by fbwoolf — APPROVED** [review]

**Pull request merged**